rules/k8s/logging-elasticsearch.rules.yaml

```diff
@@ -0,0 +1,190 @@
+  # the critical cause.
+  #
+  # Cluster health becomes YELLOW if not all index replica shards are allocated. If the goal is to have GREEN cluster
+  # health (ie number of replica shards is configured accordingly) but the status stays YELLOW for too long then this
+  # is considered serious.
+  #
+  - alert: Cluster_Health_Status_RED
+    expr: sum by (cluster) (es_cluster_status == 2)
+    for: 2m
+    labels:
+      severity: critical
+    annotations:
+      summary: 'Cluster health status is RED'
+      description: 'Cluster {{ $labels.cluster }} health status has been RED for at least 2 minutes'
+
+  - alert: Cluster_Health_Status_YELLOW
+    expr: sum by (cluster) (es_cluster_status == 1)
+    for: 20m
+    labels:
+      severity: high
+    annotations:
+      summary: 'Cluster health status is YELLOW'
+      description: 'Cluster {{ $labels.cluster }} health status has been YELLOW for at least 20 minutes'
+  #
+  # Bulk Requests Rejection
+  # =======================
+  #
+  # Sudden spikes (increases) in number of rejected bulk requests is considered serious. It means the node can not keep
+  # up with incoming bulk request indexing pace.
+  #
+  # Check https://docs.google.com/presentation/d/1X1rKozAUuF2MVc1YXElFWq9wkcWv3Axdldl8LOH9Vik/edit#slide=id.gb41e27854_0_27
+  - alert: Bulk_Requests_Rejection_HIGH
+    expr: round( bulk:reject_ratio:rate2m * 100, 0.001 ) > 5
+    for: 10m
+    labels:
+      severity: high
+    annotations:
+      summary: 'High Bulk Rejection Ratio - {{ $value }}%'
+      description: 'High Bulk Rejection Ratio at {{ $labels.node }} node in {{ $labels.cluster }} cluster'
+  #
+  # Disk Usage
+  # ==========
+  #
+  # There are two important thresholds that impact how ES node allocates index shards.
+  #   - 85% used - Low watermark
+  #   - 90% used - High watermark
+  #
+  # Disk allocation thresholds (low, high watermarks) are explained in [1,2].
+  #
+  # It is important the make sure there is enough free disk space for automatic background Lucene segment merges.
+  # Ideally as much free disk space as total sum of actual segment size (ie, data can fit into disk twice).
+  #
+  # [1] https://www.elastic.co/guide/en/elasticsearch/reference/2.4/disk-allocator.html
+  # [2] https://www.elastic.co/guide/en/elasticsearch/reference/5.5/disk-allocator.html
+  #
+  - alert: Disk_Low_Watermark_Reached
+    expr: >
+            sum by (cluster, instance, node) (
+              round(
+                (1 - (
+                  es_fs_path_available_bytes /
+                  es_fs_path_total_bytes
+                )
+              ) * 100, 0.001)
+            ) > 85
+    for: 5m
+    labels:
+      severity: alert
+    annotations:
+      summary: 'Low Watermark Reached - disk saturation is {{ $value }}%'
+      description: 'Low Watermark Reached at {{ $labels.node }} node in {{ $labels.cluster }} cluster'
+
+  - alert: Disk_High_Watermark_Reached
```

Comment:
Would it be possible to predict when this happens with predict linear? It seems like this is a situation we would like to resolve before it actually occurs.

Reply:
`Disk_High_Watermark_Reached` is quite serious and we want to avoid it for sure. I think it would be possible to use predictive functions provided by Prometheus but it will take some more time to test it properly.

The real ES disk saturation over time is more like saw chain shape and we need to test what kind of predictive function works best in our use case (depends on number of index segments, indexing activity, ... etc). The prediction formula may also be influenced by our needs like how may minutes/hours ahead we need to know this limit is approaching (btw do you know SLA react times here?).

I want to keep it simple for the first round but improving in the future is welcome (and expected 😄).

Reply:
> Prometheus but it will take some more time to test it properly.

In that case at least do a warning severity alert before it reaches this level. In general, a small amount of very good alerts is a lot better than any number of poor ones, as it otherwise easily causes alert fatigue and people start ignoring them.